### ISSUE TITLE
Before/after actions not included in the action_pack span

### ISSUE THREAD

**Issue by viraptor** [DataDog/dd-trace-rb#2100]
**Current behaviour**

Rails callbacks (before/after) are not a part of the action span.
In the `before_action` callback, active_span is:
```
SpanOperation(name:rack.request,sid:1008860920845926479,tid:1866324838581500214,pid:0)
```
But in the action call:
```
SpanOperation(name:rails.action_controller,sid:3701457545195306293,tid:1866324838581500214,pid:1008860920845926479)
```

**Expected behaviour**

active_pack spans should include the callbacks. That could be done either through a separate span or by extending the current one.

I'm raising this as a bug rather than a feature request, because otherwise we end up in situations where (for example) rack seems to own spans from `active_record`, which is not true: 
![Screen Shot 2022-06-22 at 4 08 39 pm](https://user-images.githubusercontent.com/188063/174955773-87b1c13e-ac53-405a-bc71-6f405ffc68cd.png)
Those queries are made in the `before_action` on the controller.

**Steps to reproduce**

```
class SomeController < ApplicationController
  before_action :bar

  def foo
    Datadog::Tracing.trace("this.goes.in.controller") {}
  end

  private
  
  def bar
    Datadog::Tracing.trace("this.goes.in.rack") {}
  end
end
```

**Environment**

* **ddtrace version:** 1.1.0
* **Configuration block (`Datadog.configure ...`):**
```
    c.tracing.instrument :rails, service_name: 'marketplace'
    c.tracing.instrument :rack, request_queuing: true, web_service_name: 'marketplace-rack'
```
* **Ruby version:** 2.7
* **Operating system:** linux
* **Relevant library versions:**

**Comment by delner** [comment]
ActionPack instrumentation doesn't cover `before/after_action`; this is known behavior. We wanted to add coverage for this, but the nature of how it is implemented in Rails (as an ActiveSupport notification) makes it incredibly difficult. See this ancient PR: https://github.com/DataDog/dd-trace-rb/pull/333

**Comment by viraptor** [comment]
So I've got an ApiController base class which knows how to name the resource correctly, but doesn't actually implement any endpoint. Only controllers inheriting from it actually have the actions and I don't want to replicate the resource naming in them.
The problem is that the `Class#method` naming in the API classes doesn't really map nicely to the requests (either being too general or too specific), so I want to use a custom name that depends on request parameters. 

I can't (as far as I know) use the span processor / filter here, because I don't have access to the parameters anymore.

After some playing around last week, I ended up with renaming the trace instead. Specifically this works: (after so that action_pack doesn't override it)

```
after_action :set_tracing_transaction_name
def set_tracing_transaction_name
   ...
   Datadog::Tracing.active_trace&.resource = "ApiChunk:#{transaction_name}"
end
```

So at this point I'd like to rename the controller span, but it became more of a "nice to have". I can live with only the trace being renamed.

**Comment by KJTsanaktsidis** [comment]
Another fun side effect of this is that if a `before_action` renders or raises, it never hits the actionpack instrumentation (because it's patched into `ActionController::Metal`, and it's therefore lower in the ancestor chain of the controller than `AbstractController::Callbacks` which gets mixed into `ActionController::Base` (which is a subclass of `Metal`).

i.e. the ancestor chain looks like...

```
MyFunController
ActionController::Base # We inherit from this
AbstractController::Callbacks # This is included into ActionController::Base
Datadog::Tracing::Contrib::ActionPack::ActionController::Metal # This is prepended onto ActionController::Metal
ActionController::Meta
```

When `process_action` gets called, the before callbacks get run in `AbstractController::Callbacks` before it calls super, and so if it renders or raises, `Datadog::Tracing::Contrib::ActionPack::ActionController::Metal#process_action` is never called.

I mostly care about this because the actionpack instrumentation is responsible for setting the resource name of the top-level Rack span to be the controller name. This means we wind up with a mountain of traces simply called `Zendesk::Application#GET`, which is not terribly helpful (that happens because the Rack instrumentation [patches the middleware name into all of our middlewares](https://github.com/dataDog/dd-trace-rb/blob/a364d8b5b916fe0593144b780bdbeb7c917623f7/lib/datadog/tracing/contrib/rack/patcher.rb#L52) and our Rails application is the final "Middleware" from Rack's point of view.

**Comment by delner** [comment]
@KJTsanaktsidis Right, because its the `ActionController` instrumentation that's updating the trace resource with a controller name hasn't run yet before you're raising an error.

In theory, this could be done at a routing level if we instrumented the routing behavior. That would always run before any controllers, but should be able to show a route name.

**Comment by KJTsanaktsidis** [comment]
Yeah instrumenting the router for this would make sense actually

**Comment by delner** [comment]
@KJTsanaktsidis Cool. It's a bit separate from "instrument before/after actions", so we could open that as its own feature request.

**Comment by beauraF** [comment]
Hello everyone 👋

I'm bringing this up again because the two problems described here are two major issues for us:
1. before/after action not included in the `rails.action_controller` span
2. missing resource name on action that stop in middleware or before_action

For 2, I have open a dedicated [issue](https://github.com/DataDog/dd-trace-rb/issues/3482) for it. I'm sorry @delner I opened it as a "bug report", and just saw you proposed to open it as a "feature request".

For 1, I'll be looking this week to see if it's possible to push upstream instrumentation around before/around/after action. That should make thing easier?

**Comment by sdhull** [comment]
Wanted to chime in here to add my support for instrumenting at the routing level.

IMO the fact that requests to an endpoint will not be counted in APM if a `before_action` raises or renders should be considered a P0 bug, because it erodes trust in the APM product.

Let me tell you a story. Today I was making statements to my colleagues like "this endpoint only gets X requests / min, look at APM" and one of my colleagues proceeds to send me a DD logs query showing actual volume to that endpoint is >100x what APM shows. I don't know about you, but to me missing a request or two is not that big a deal (typically). But being off by an order (or two) of magnitude is a _big deal_.

Thanks for coming to my TED Talk

**Comment by trevarmand** [comment]
+1. Being able to use before_actions to add custom tags to spans is much simpler than having to repeat code everywhere. In our prior APM offering we were able to define a before_action in our ApplicationController (parent to all controllers in the app) that could add attributes to every endpoint via a couple lines of code in a single file.
